service-wfs-client/src/main/java/org/oskari/service/wfs/client/OskariWFSClient.java

```diff
@@ -61,17 +65,63 @@ protected static String getBBOX(ReferencedEnvelope bbox) {
                 srsName);
     }
 
-    protected static byte[] getResponse(String endPoint,
-                                        String user, String pass, Map<String, String> query) {
+    protected static SimpleFeatureCollection getFeatures(String endPoint,
+            String user, String pass, Map<String, String> query,
+            CoordinateReferenceSystem crs, boolean tryGeoJSON, OskariGMLDecoder gmlDecoder) {
+        String url; // for debugging
+
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        byte[] response;
+        Map<String, String> responseHeaders;
+        SimpleFeatureCollection fc;
+
+        if (tryGeoJSON) {
+            // First try GeoJSON
+            query.put("OUTPUTFORMAT", "application/json");
+            url = IOHelper.constructUrl(endPoint, query);
+            responseHeaders = OskariWFSClient.readResponseTo(endPoint, user, pass, query, baos);
+            response = baos.toByteArray();
+            // TODO: Select parsing algorithm based on response headers (Content-Type)
+            fc = parseGeoJSON(response, crs, url);
+            if (fc != null) {
+                return fc;
+            }
+            // Try to parse the same response as GML
+            fc = parseGML(response, crs, url, user, pass, gmlDecoder);
+            if (fc != null) {
```

Comment:
Add:

LOG.warn("Requested JSON but got GML. Possibly misconfigured service for", url); or something similar.

Reply:
Added